### PR TITLE
stdenv: mesonFlags: specify more targetPrefix'd binaries

### PR DESCRIPTION
**Pull request by Artturin**
see commit msgs

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by alyssais** [comment]
Can you please explain in the commit message what the purpose of this change is?  "improve" isn't really enough information and even the linked PR doesn't really have the context.

File: pkgs/stdenv/generic/make-derivation.nix [hunk 1]
```diff
@@ -394,6 +394,20 @@ else let
 
             [binaries]
             llvm-config = 'llvm-config-native'
+
+            ${lib.optionalString (stdenv.cc ? bintools && stdenv.cc.bintools.targetPrefix != "") ''
```
Comment (alyssais) [inline pkgs/stdenv/generic/make-derivation.nix:398]: Do we need to check `stdenv.cc.bintools.targetPrefix != ""`?

Comment (Artturin) [reply pkgs/stdenv/generic/make-derivation.nix]: not really, i just wanted to keep the meson file small but on second thought its better to keep the behaviour the same.

**Comment by Artturin** [comment]
TODO: investigate adding using the env vars we already set by adding entries to 
https://github.com/mesonbuild/meson/blob/1e1dee738b7eae9e9c22601522798f4945391379/mesonbuild/envconfig.py#L114

**Comment by Artturin** [comment]
https://github.com/mesonbuild/meson/pull/10640

**Comment by alyssais** [comment]
The commit message is fantastic now, thank you! :)  (And thanks for going upstream!)

File: pkgs/development/tools/build-managers/meson/default.nix [hunk 2]
```diff
@@ -29,6 +29,11 @@ python3.pkgs.buildPythonApplication rec {
     # https://github.com/mesonbuild/meson/pull/6827
     ./more-env-vars.patch
 
+    # Use more binutils variables, so we don't have to define them in stdenv.
+    # pr has been merged
+    # https://github.com/mesonbuild/meson/pull/10640
+    ./more-binutils-vars.patch
```
Comment (SuperSandro2000) [inline pkgs/development/tools/build-managers/meson/default.nix:35]: why are we not using fetchpatch?

Comment (Artturin) [reply pkgs/development/tools/build-managers/meson/default.nix]: switched

**Review by Ericson2314 — APPROVED** [review]
This is really great! Great job!!

**Comment by alyssais** [comment]
This broke `pkgsStatic.glib`:

```
FAILED: glib/tests/getpwuid-preload.so
x86_64-unknown-linux-musl-gcc  -o glib/tests/getpwuid-preload.so glib/tests/getpwuid-preload.so.p/getpwuid-preload.c.o -Wl,--as-needed -Wl,--no-undefined -shared -fPIC -Wl,--start-group -Wl,-soname,getpwuid-preload.so -Wl,--end-group
/nix/store/cja158bhs3cplv0w8zdib822524rwqw9-x86_64-unknown-linux-musl-binutils-2.38/bin/x86_64-unknown-linux-musl-ld: /nix/store/vn93aipraqpva34gz3y9vzjfw3kzdyac-x86_64-unknown-linux-musl-stage-final-gcc-11.3.0/lib/gcc/x86_64-unknown-linux-musl/11.3.0/crtbeginT.o: relocation R_X86_64_32 against hidden symbol `__TMC_END__' can not be used when making a shared object
/nix/store/cja158bhs3cplv0w8zdib822524rwqw9-x86_64-unknown-linux-musl-binutils-2.38/bin/x86_64-unknown-linux-musl-ld: failed to set dynamic section sizes: bad value
collect2: error: ld returned 1 exit status
```

**Comment by alyssais** [comment]
Ah, because now glib builds its tests, and lots of those tests involve building shared libraries.  Maybe I can get them to disable the tests if default_library=static.

**Pull request merged**